### PR TITLE
Optionally email the admin if a send task failed and we suspend the source cleanup

### PR DESCRIPTION
**Pull request by jimklimov**
Should send the error summary to specified recipient(s) and fix the issue #499 as far as I (original poster) am concerned :) 

Tested on a local system, delivers to one or two (comma-separated) local mailboxes using the standard `/usr/lib/sendmail` or `/usr/sbin/sendmail` alias (exim, postfix also provide that interface). Internet delivery is up to MTA setup and outside znapzend's interest.

````
:; znapzend --noaction --runonce=nvpool/SHARED/var/test --inherited --debug --cleanOffline --mailErrorSummaryTo=root,jim

...
[2020-08-20 00:29:08.80293] [26384] [warn] ERROR: 1 send task(s) below failed for nvpool/SHARED/var/test, but "cleanOffline" mode is on, so proceeding to clean up source dataset carefully:
[2020-08-20 00:29:08.80315] [26384] [warn]  +-->   destination 'backup-adata/snapshots/nvpool/SHARED/var/test' does not exist or is offline. ignoring it for this round...
[2020-08-20 00:29:08.80643] [26384] [warn] Sending a copy of the report above to root,jim
[2020-08-20 00:29:08.83905] [26384] [debug] checking to clean up snapshots recursively from source nvpool/SHARED/var/test
...
````

If the mail program does not exist in standard path (currently hardcoded, can be made an argument later), it is noted but not fatal:
````
[2020-08-20 00:27:28.54081] [26359] [warn] ERROR: 1 send task(s) below failed for nvpool/SHARED/var/test, but "cleanOffline" mode is on, so proceeding to clean up source dataset carefully:
[2020-08-20 00:27:28.54089] [26359] [warn]  +-->   destination 'backup-adata/snapshots/nvpool/SHARED/var/test' does not exist or is offline. ignoring it for this round...
Can't exec "/usr/sbin/sendmailsss": No such file or directory at /export/home/jim/shared/znapzend/bin/../lib/ZnapZend.pm line 786.
Can't open /usr/sbin/sendmailsss to send a copy of the report above to root,jim!
````

Message in the box:
````
From jim@jimoi.local Thu Aug 20 00:29:08 2020
Return-Path: <jim@jimoi.local>
Received: from jimoi.local (jimoi [127.0.0.1])
        by jimoi.local (8.15.2+Sun/8.15.2) with ESMTP id 07JMT880026388;
        Thu, 20 Aug 2020 00:29:08 +0200 (CEST)
Received: (from jim@localhost)
        by jimoi.local (8.15.2+Sun/8.15.2/Submit) id 07JMT8wN026387;
        Thu, 20 Aug 2020 00:29:08 +0200 (CEST)
Date: Thu, 20 Aug 2020 00:29:08 +0200 (CEST)
From: jim <jim@jimoi.local>
Message-Id: <202008192229.07JMT8wN026387@jimoi.local>
To: root@jimoi.local, jim@jimoi.local
Subject: znapzend replication error summary
Content-Length: 290

-------
ERROR: 1 send task(s) below failed for nvpool/SHARED/var/test, but "cleanOffline" mode is on, so proceeding to clean up source dataset carefully:
 +-->   destination 'backup-adata/snapshots/nvpool/SHARED/var/test' does not exist or is offline. ignoring it for this round...
-------
````

### CODE REVIEW

File: lib/ZnapZend.pm [hunk 1]
```diff
@@ -53,6 +53,7 @@ has nodelay                 => sub { 0 };
 has skipOnPreSnapCmdFail    => sub { 0 };
 has skipOnPreSendCmdFail    => sub { 0 };
 has cleanOffline            => sub { 0 };
+has mailErrorSummaryTo      => sub { q{} };
```
Comment (oetiker) [inline lib/ZnapZend.pm:56]:
```suggestion
has 'mailErrorSummaryTo';
```

File: lib/ZnapZend.pm [hunk 2]
```diff
@@ -752,14 +753,53 @@ my $sendRecvCleanup = sub {
...
+            $errline = 'ERROR: ' . scalar(@sendFailed) . ' send task(s) below ' .
+                'failed for ' . $backupSet->{src} . ', but "cleanOffline" mode ' .
+                'is on, so proceeding to clean up source dataset carefully:' ;
+        } else {
+            $errline = 'ERROR: suspending cleanup source dataset ' .
+                $backupSet->{src} . ' because ' .
+                scalar(@sendFailed) . ' send task(s) failed:' ;
+        }
+        $self->zLog->warn($errline);
+        if ($self->mailErrorSummaryTo ne '') {
```
Comment (oetiker) [inline lib/ZnapZend.pm:772]:
```suggestion
        if ($self->mailErrorSummaryTo) {
```

File: lib/ZnapZend.pm [hunk 3]
```diff
@@ -752,14 +753,53 @@ my $sendRecvCleanup = sub {
...
+            $errmsg = $errline . "\n";
+        }
+        for $errline (@sendFailed) {
+            $self->zLog->warn(' +-->   ' . $errline);
+            if ($self->mailErrorSummaryTo ne '') {
+                $errmsg .= ' +-->   ' . $errline . "\n";
+            }
+        }
+
+        if ($self->mailErrorSummaryTo ne '') {
```
Comment (oetiker) [inline lib/ZnapZend.pm:782]:
```suggestion
        if ($self->mailErrorSummaryTo) {
```

**Review by oetiker — CHANGES_REQUESTED** [review]
Setting to an empty string and then comparing to it seems rather non-ideomatic ... and since a valid value will never be '0' there should be no problem with just using the value directly ...

File: lib/ZnapZend.pm (hunk 1 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:56]: Good point; so far I was just following the precedent in nearby lines :)

**Review by oetiker — CHANGES_REQUESTED** [review]

**Comment by jimklimov** [comment]
Bump? This change is in our production for a month now :)

**Review by oetiker — APPROVED** [review]
oops missed that

**Pull request merged**